src/compiler/tsbuild.ts

```diff
@@ -296,9 +297,9 @@ namespace ts {
     }
 
     export interface SolutionBuilder<T extends BuilderProgram> {
-        build(project?: string, cancellationToken?: CancellationToken): ExitStatus;
+        buildAsync(project?: string, cancellationToken?: CancellationToken): Promise<ExitStatus>;

```

Comment:
Can I break this public API?  If not, what's a clean way to have both?

Reply:
cc @sheetalkamat

Reply:
You cant break the api but you can add another menthod buildAsync instead?
You would also need to create
` getNextInvalidatedProjectAsync(cancellationToken?: CancellationToken): InvalidatedProjectAsync<T> | undefined;` as well? Because the emit is actually on invalidated project

Reply:
Do I mark the old API as deprecated or something?  We won't have any callers after the change.  Also, is there some way to share code between them?  In C#, I would just `WaitAndGetResult()` to switch from async to sync.

Reply:
There's no equivalent to `WaitAndGetResult()` in JS because the language is single-threaded; any attempt to block on a promise blocks the event loop and will basically cause a deadlock.  Once you're async, there's no way out.

Reply:
@fatcerberus so, if you want to provide both sync and async versions of an API you have to duplicate all the code?

Reply:
You can factor out the common sync parts to be called by both versions, but yes, in general you have to write the function twice.  At least, when using the `async`-`await` syntax.  This isn't usually a problem in the real world because the async implementation of an API tends to be pretty different from the sync version whenever both are present.  `await` does the blur the line a bit though.